tlatools/src/tla2sany/configuration/ConfigurationTokenManager.java

```diff
@@ -1024,9 +1024,7 @@ static public void ReInit(SimpleCharStream stream)
 {
    jjmatchedPos = jjnewStateCnt = 0;
    curLexState = defaultLexState;
-   // Bug: The following line used to read "input_stream = stream; but this bug was corrected
-   // by DRJ.  The bug should also be corrected in the .jcc file, but isn't.
-   input_stream = null;
+   input_stream = stream;
```

Comment:
Please clarify, the bug has not been fixed in the .jcc file?  If not, why?

Reply:
It is good that you have asked for that.
I have thought about that and I'll try to explain it:

With a static parser generated with JavaCC you have a static object,
which is called Configuration here. If you use that static parser on multiple input streams,
then you have to call one of the ReInit() methods before each run,
which get the new input stream as an argument.
That means you need to know, at this point of time, the stream which should be parsed.

The current implementation woks like this:
The only object which uses the Configuration object is in tla2sany/drivers/SANY.java,
which calls frontEndMain() which calls frontEndInitialize() for each of maybe multiple TLA-files.
In frontEntInitialize(), which is the only method which uses Configuration,
there is first a call to the Configuration.ReInit() and then a call to Configuration.load().

This ReInit() method from DRJ is mostly a copy of the one provided by JavaCC,
but it has no parameters and it is altered that way that it uses the static member input,
which is not defined in the first run. That is an error in the current code.
Notice that the stream of the input is not known at this point of time.
Also it calls the ReInit() methods of ASCII_CharStream/SimpleCharStream and ConfigurationTokenManager

Then Configuration.load() is called, which:

- determinates the source of input (the String defaultconfig or the File "config.src"),
  creates a Stream and saves that in the static member input,
- instantiates a new !!! static !!! parser with the correct input stream,
- parses that stream

If there are multiple TLA files, then the static member input contains the input Stream of the run before, which is then used in the next call to Configuration.ReInit().

**Conclusion**:
I hope you see that this handwritten ReInit() method from DRJ is the source of these bugs,
which than required that fixes in the other files.

BTW, in the TLA-parser there is also the same code in SimpleCharStream and TLAplusParserTokenManager which has worked for the same time and did not need the bugfixes there.

**How to proceed**:
If you accept all patches until the ReInit() removal patches,
then I would rebase them, so you can pull these.
Then I would make another branch for the patches which will result in a sane static parser,
the way it is thought by the JavaCC developers.
After that the transition from static to non-static can be done seperately.